File: post_receive.lua
#!/usr/bin/lua

local json = false
local JSON = false
pcall(function() JSON = require "JSON" end) -- JSON: http://regex.info/code/JSON.lua
pcall(function() json = require "json" end) -- LuaJSON, if available
local socket = require "socket"
local lfs = false
pcall(function() lfs = require "lfs" end)

local old = arg[1]
local new = arg[2]
local ref = arg[3] or "(no ref)"

if not old or not new then
    print("Usage: post-receive [old] [new] [ref]")
    os.exit()
end

local commit = {repository="git", ref=ref}

local pwd = ""
if lfs then
    pwd = lfs.currentdir()
else
    local p = io.popen("pwd")
    if p then
        pwd = p:read("*l")
        p:close()
    end
end
local project = pwd:match("([^/]+)$") or "unknown"

local pipe = io.popen("git show --name-only " .. new)
commit.hash = new
commit.shortHash = commit.hash:sub(1,7)
commit.project = project

while true do
    local line = pipe:read("*l")
    if not line or line:len() == 0 then break end
    local key, val = line:match("^(%S+):%s+(.+)$")
    if key and val then
        key = key:lower()
        if key == "author" then
            local author, email = val:match("^(.-) <(.-)>$")
            commit.author = author or "(uknown)"
            commit.email = email or "(unknown)"
        else
            commit[key] = val
        end
    end
end

commit.log = ""
while true do
    local line = pipe:read("*l")
    if not line or line:len() == 0 then break end
    commit.log = commit.log .. line:gsub("^    ", "") .. "\n"
end

commit.log = commit.log:gsub("\n$", "")


commit.files = {}
while true do
    local line = pipe:read("*l")
    if not line or line:len() == 0 then break end
    table.insert(commit.files, line)
end

pipe:close()

if #commit.files == 1 then
    commit.changes = commit.files[1]
else
    commit.changes = #commit.files .. " files"
end


local output = ""
if JSON then 
    output = JSON:encode({commit=commit})
elseif json then
    output = json.encode({commit=commit})
end
print(output)
local s = socket.tcp()
s:settimeout(0.5)
local success, err = s:connect("127.0.0.1", 2069)
if not success then
    os.exit()
end

s:send("POST /json HTTP/1.1\r\n")
s:send("Host: localhost\r\n\r\n")
s:send(output .."\r\n")
s:shutdown()
s:close()

File: post_receive.py
#!/usr/bin/env python

import re
import os
import sys
import json
import ConfigParser;
import sys
import logging
from subprocess import Popen, PIPE
from httplib2 import Http

config = ConfigParser.ConfigParser()
config.read('gitpubsub.cfg')

logging.basicConfig(filename='gitpubsub.log',level=logging.DEBUG)

# Set this to point to your local gitpubsub server
# Since this requires a gitpubsub.cfg in the local 
# directory, we'll also hardcode it for now.
postURL = "http://localhost:2069/json"
try:
    postURL = config.get("Server", "URL", 0)
except:
    logging.info("gitpubsub.cfg not found, using hardcoded defaults")
    pass

project = os.path.basename(os.getcwd())

while True:
    line = sys.stdin.readline()
    if not line:
        break
    [old, new, ref] = line.split(None, 2)
    logging.info("Posting commit message for ref %s in project %s", new, project)

    process = Popen(["git", "show", "--name-only", new], stdout=PIPE)
    exit_code = os.waitpid(process.pid, 0)
    output = process.communicate()[0]

    commit = {'ref': ref, 'repository': "git", 'hash': new, 'project': project}

    arr = output.split("\n\n", 3)
    headers = arr[0]
    log = arr[1]
    if len(arr) == 3:
        commit['files'] = re.findall(r"([^\r\n]+)", arr[2])

    parsed = dict(re.findall(r"(?P<name>[^:\n]+): (?P<value>[^\r\n]+)", headers))

    author = re.match(r"^(.+) <(.+)>$", parsed.get("Author", "?? <??@??>"))
    if author:
        commit['author'] = author.group(1)
        commit['email'] = author.group(2)
    else:
        commit['author'] = "Unknown"
        commit['email'] = "unknown@unknown"

    data = json.dumps({'commit': commit}) + "\n\n"
    logging.info("Posting commit data to URL %s", postURL)
    try:
        Http().request(postURL, "PUT", data)
        logging.info("Commit %s was posted", new)
    except:
        logging.warning("Could not connect to %s, commit not reported", postURL)
        pass

